### ISSUE TITLE
Port Rubi 4.16 to Symja

### ISSUE THREAD

**Issue by axkr** [axkr/symja_android_library#70]
Port Rubi 4.16 to Symja

The converter will be maintained here: [/symja_android_library/tools/src/main/java/org/matheclipse/core/rubi](https://github.com/axkr/symja_android_library/tree/master/symja_android_library/tools/src/main/java/org/matheclipse/core/rubi)
(files ConvertRubi.java, ConvertRubiUtilityFunctions.java)

See: 
* https://github.com/axkr/symja_android_library/wiki/Porting-Rubi-Integration-rules-to-Symja
* https://github.com/RuleBasedIntegration/Rubi/wiki

ToDo:

- [X] ~~create new branch rubi416~~ switched back to master
- [X] fix StackOverflowError in  `org.matheclipse.core.system.CompleteTestSuite`  JUnit tests
- [X] fix unintegrable() function errors in `org.matheclipse.core.system.CompleteTestSuite`  JUnit tests (solved by filtering and don't using "Unintegrable" right-hand-side rules
- [X] define an own "priority" for each rule by simply counting up by +1 (see ConvertRubi.GLOBAL_COUNTER)
- [X] improve speed for `Integrate` function ( a881780 )
- [X] create new Rubi JUnit tests (with timeout setting) ( 4843200 )
- [X] easier Junit tests: allow input of the form `Integrate::PolyQ[... ,x ]` with leading `Integrate::` identifier prefix
- [X] small improvement for loading performance by `IIntegrate()` function specialiced to lhs `Int()` rules
- [X] convert these `SetDelayed` structures [RubiRules4.16.0_FullLHS.m](https://github.com/axkr/Rubi-1/blob/master/Rubi/RubiRules4.16.0_FullLHS.m) and these [Rubi utility functions](https://github.com/axkr/Rubi-1/blob/master/Rubi/IntegrationUtilityFunctions.m) into internal [Symja Java Rubi rules](https://github.com/axkr/symja_android_library/tree/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/integrate/rubi). The priority set for the `Int[]` functions should correspond with the line number of the input file.
- [X] added [Wiki page](https://github.com/axkr/symja_android_library/wiki/Porting-Rubi-Integration-rules-to-Symja)
- [ ] fix more bugs in the JUnit Rubi tests suite  
  * [ ] [AlgebraicFunctions](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/AlgebraicFunctions.java) 1329 Test / 1 Failures 
  * [X] [Exponentials](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/Exponentials.java) 195 Tests / 0 Failures
  * [X] [HyperbolicFunctions](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/HyperbolicFunctions.java) 116 Tests / 0 Failures
  * [X] [IndependentTestSuites](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/IndependentTestSuites.java) 279 Tests / 0 Failures
  * [X] [InverseHyperbolicFunctions](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/InverseHyperbolicFunctions.java) 100 Tests / 0 Failures
  * [X] [InverseTrigFunctions](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/InverseTrigFunctions.java)  66 Tests / 0 Failures
  * [X] [Logarithms](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/Logarithms.java) 82 Tests / 0 Failures
  * [X] [SpecialFunctions](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/SpecialFunctions.java) 260 Tests / 0 Failures  
  * [ ] [TrigFunctions](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/test/java/org/matheclipse/core/rubi/TrigFunctions.java)  531 Tests / 4 Failures

**Comment by tranleduy2000** [comment]
I am tested on Android device from 4.0 to 9.0, some test can not calculate integrate although I increase stack size up to 4MB

**Comment by tranleduy2000** [comment]
```Java
org.matheclipse.core.system.MainTestCase#testSystem172
org.matheclipse.core.system.RubiIntegrationTest#testRationalFunction001
org.matheclipse.core.system.RubiIntegrationTest#testRationalFunction002
org.matheclipse.core.system.RubiIntegrationTest#testRationalFunction003

junit.framework.AssertionFailedError: com.google.common.util.concurrent.ExecutionError: java.lang.StackOverflowError
```

```Java
org.matheclipse.core.system.RubiIntegrationTest#testArcSin
org.matheclipse.core.system.RubiIntegrationTest#testIssue84

junit.framework.AssertionFailedError: java.lang.StackOverflowError
````

```Java
```

```Java
org.matheclipse.core.system.RubiIntegrationTest#testTrig001
junit.framework.AssertionFailedError: expected:<[-((E^(I*2*a+I*2*b*x)*(1-I*2*b*x))/(8*b^2)+(1+I*2*b*x)/(8*b^2*E^(I*2*a+I*2*b*x)))/
2+x^2/4]> but was:<[x^2/4-(Cos(2*a+2*b*x)/(4*b^2)+(x*Sin(2*a+2*b*x))/(2*b))/2]>
```


Test success on Windows but failed on Android.
```Java
org.matheclipse.core.system.SerializableTest#testPowerSeries
org.matheclipse.parser.client.SyntaxError: Syntax error in line: 1 - ')' expected.
Series(Log(x),{x,a,4})
^
```

**Comment by tranleduy2000** [comment]
``java.lang.NullPointerException`` occur when write serializable ``IExpr``.

```
org.matheclipse.core.expression.Symbol.equals (Unknown Source:41)
java.util.HashMap.getNode (HashMap.java:571)
java.util.HashMap.get (HashMap.java:556)
org.matheclipse.core.expression.AST0.writeExternal (Unknown Source:46)
```

**Comment by axkr** [comment]
Commited 5e1ac56 - create new EvalEngine() for every JUnit Rubi test

**Comment by axkr** [comment]
The current implementation will be released in 2.0.0. Please open new issue if necessary.
